pdfrw/objects/pdfstring.py

```diff
@@ -13,62 +77,83 @@ class PdfString(str):
+    escapes = {
+        b'\n': b'\\n',
+        b'\r': b'\\r',
+        b'\t': b'\\t',
+        b'\b': b'\\b',
+        b'\f': b'\\f',
+        b'(' : b'\\(',
+        b')' : b'\\)',
+        b'\\': b'\\\\'
+    }
+
+    unescapes = {v[1]: k for k, v in escapes.items()}
+
+    # Possible string escapes from the spec:
+    # 1. \[nrtbf\()]: simple escapes
+    # 2. \\d{1,3}: octal. Must be zero-padded to 3 digits if followed by digit
+    # 3. \<end of line>: line continuation. We don't know the EOL marker
+    # used in the PDF, so accept \r, \n, and \r\n.
+    unescape_re = re.compile(br'\\(([nrtbf\(\)\\])|(\d{1,3})|(\r|\r\n|\n))')
+
+    def decode_escaped(self):
+        def handle_escape(m):
+            if m.group(2):  # direct escape
+                return self.unescapes[m.group(2).decode('ascii')]
+            if m.group(3):  # octal
+                value = int(m.group(3).decode('ascii'), 8)
+                return bytes(bytearray([value]))
+            if m.group(4):  # line continuation
+                return b''
+            # The PDF spec says that if the backslash doesn't form a valid
+            # escape, the backslash is ignored.
+            return m.group(1)
+
+        return self.unescape_re.sub(handle_escape, bytearray(self[1:-1]))
 
+    def decode_bytes(self, bytestr):
+        if bytestr[:2] == codecs.BOM_UTF16_BE:
+            return bytestr[2:].decode('utf-16-be')
         else:
-            return self.decode_hex(remap, twobytes)
-
-    def encode(cls, source, usehex=False):
-        assert not usehex, "Not supported yet"
-        source = source.replace('\\', '\\\\')
-        source = source.replace('(', '\\(')
-        source = source.replace(')', '\\)')
-        return cls('(' + source + ')')
-    encode = classmethod(encode)
+            return bytestr.decode('pdfdocencoding')
+
+    def decode_hex(self):
+        hexstr = self[1:-1]
+        if len(hexstr) % 1: # odd number of chars indicates a truncated 0
+            hexstr += '0'
+        return binascii.unhexlify(hexstr)
+
+    def decode(self):
+        if self.startswith('(') and self.endswith(')'):
+            raw = self.decode_escaped()
+
+        elif self.startswith('<') and self.endswith('>'):
+            raw = self.decode_hex()
+
+        else:
+            raise ValueError('Invalid PDF string "%s"' % repr(self))
+
+        return self.decode_bytes(raw)
+
+    escape_re = re.compile(br'(([^\x20-\x7e])|([\n\r\t\b\f\(\)\\]))')
+
+    @classmethod
+    def encode(cls, source):
+        def handle_char(m):
+            if m.group(1):  # unprintables
+                return ('\\%03o' % ord(m.group(1))).encode()
+            if m.group(2):  # simple escapes
+                return escapes[m.group(2)].encode()
+
+        try:
+            raw = source.encode('pdfdocencoding')
+            escaped = cls.escape_re.sub(handle_char, raw)
+
+            return cls('(' + escaped.decode('ascii') + ')')
+
+        except ValueError:
```

Comment:
I assume this is to catch when the characters in the string fall outside the set defined for pdfdocencoding. If that's the case, can we catch the more specific `UnicodeEncodeError`, rather than catching all ValueErrors?